New article "Enable App Actions on Windows for PWA" (#3455)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-chromium/how-to/app-actions.md

Author: consider redesigning the first few headings per standard conventions eg no "Intro" heading.

Author: check whether can avoid h5 headings; flatten outline to h2/h4/h6.

Author: check the wording: 3 levels of formal product name:
App Actions on Windows - 1st occurr.
App Actions
actions -

---

> title: Enable App Actions on Windows for PWA

Author: ```suggestion
title: Enable App Actions on Windows in your PWA
```
global change title

      * [How to package store PWA in Edge Canary](#how-to-package-store-pwa-in-edge-canary)
   * [Testing your PWA in Windows](#testing-your-pwa-in-windows)
      * [Using Windows actions test tool](#using-windows-actions-test-tool)

<!-- ====================================================================== -->
## Introduction

Here's how to author and test an app action for the App Actions on Windows framework, for a Progressive Web App (PWA) that you've created.

<!-- ------------------------------ -->
#### What's App Actions on Windows

An App Action on Windows is an individual unit of behavior that a Windows app can implement and register.  The App Action can then be accessed from other apps and experiences, seamlessly integrating into user workflows. To learn more about App Actions on Windows, see [App Actions on Windows](https://learn.microsoft.com/en-us/windows/ai/app-actions/).

An App Action is an atomic unit of functionality.  An app builds and registers actions, and then Windows or other apps can recommend registered actions to the user at contextually relevant times and locations within the user workflow.

<!-- ------------------------------ -->

> #### Why should you build actions for your app

Author: deleted heading, moved content up to lead-in

> App developers can build actions to increase discoverability and engagement for their features with app-to-app and agentic interactions.

Author: ```suggestion
You can build actions to increase discoverability and engagement of your app's features by using app-to-app and agentic interactions.  The app actions framework makes it possible for you to integrate your app with other apps or with AI agents.
```

The Windows App Actions framework is for exposing and executing key pieces of app functionality – it is not for generic user activities like opening context menus or right-clicking.

See [Recommended scenarios for App Actions](/windows/ai/app-actions#recommended-scenarios-for-app-actions) in _App Actions on Windows Overview_.

<!-- ====================================================================== -->
## Get started

<!-- ------------------------------ -->
#### Define the actions your PWA supports

This guide illustrates how to author an action using one of the sample PWAs, [wami, the Web App to Manipulate Images](https://github.com/MicrosoftEdge/Demos/tree/main/wami).  The Wami demo supports image manipulation tasks, such as resizing or applying post-processing styles.  The following example demonstrates creating a **Paint** action, which generates an oil-painting-style version of the image.

The **Paint** action accepts the image file that user selected and returns the stylized output.  The **Paint** action is defined in a descriptive file named `ActionsManifest.json`, per the next section.

<!-- -------------------- -->

> ##### Prepare ActionsManifest.json

Author: flatten toc eg remove this heading.
rendered:
https://github.com/mikehoffms/edge-developer/blob/user/mikehoffms/branch-02/microsoft-edge/progressive-web-apps-chromium/how-to/app-actions.md#prepare-actionsmanifestjson

      }]
   }]
}
```

<!-- ---------- -->
###### Understanding the fields in ActionsManifest

For the definition of each JSON property, see [Action definition JSON properties](/windows/ai/app-actions/actions-json#action-definition-json-properties) in _Action definition JSON schema for App Actions on Windows_.

Under `Invocation`, only the `uri` type is supported.  The `com` type is not applicable for PWAs.

<!-- ---------- -->
###### Supported Entity types for PWA

The `ActionEntityKind` enumeration specifies the types of entities that are supported by App Actions on Windows.  In the context of a JSON action definition, the entity kinds are string literals that are case-sensitive.  The supported entities are listed in [ActionEntityKind enumeration](/windows/ai/app-actions/actions-json#actionentitykind-enumeration) in _Action definition JSON schema for App Actions on Windows_.

`StreamingText` and `RemoteFile` are not yet supported for PWA.  For `File`, `Photo`, `Document` and `Text` entity kinds, see [Entity properties](/windows/ai/app-actions/actions-json#entity-properties) in _Action definition JSON schema for App Actions on Windows_.

<!-- -------------------- -->

> ##### How does Windows App Actions framework call App Actions in a PWA?

Reviewer: ```suggestion
##### How does Windows App Actions framework call App Actions in a PWA
```

<!-- ---------- -->
###### PWA launch through URI Activation

A PWA action is launched by Windows through custom protocol URI activation.  A PWA must register for custom protocol handling with the `protocol_handlers` field in the web app manifest.  Windows' `LaunchUriAsync(Uri, LauncherOptions, ValueSet)` method is used by the Action Framework for URI activation.

`LaunchUriAsync` doesn't return a value or error status.

<!-- ---------- -->
###### URI scheme

A custom protocol URI should be used, not an HTTPS URI.

For PWAs with many actions, use syntax:

```json
"uri": "web+{app name}://{custom action name}"
```

For example:

```json
"uri": "web+wami://paint"
```

<!-- ---------- -->

> ###### Receiving ValueSet data from Windows and mapping it to Share Target

Reviewer: ```suggestion
###### Receive ValueSet data from Windows and map it to Share Target
```

}]
```

<!-- ---------- -->
###### Web Manifest validation for `share_target`

The `share_target` field must be present in the manifest:

* The method must be "POST".
* The `enctype` must be `multipart/form-data`.
* `title`, `text` and `url` in `params` are required.  `files` is optional if you don't want your PWA to support the normal _share target_ feature. Please see [Receiving shared content](https://learn.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/share#receiving-shared-content) for more info. 

Example `share_target` declaration:

```json
"share_target": {
   "action": "./share-target",
   "method": "POST",
   "enctype": "multipart/form-data",
   "params": {
      "title": "title",
      "text": "text",
      "url": "url"
   }
}
```

<!-- ---------- -->
###### Web Manifest validation for `launch_handler`

You can set [launch_handler](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/launch_handler) as `navigate-existing` into your PWA manifest, if you want to keep a single PWA window.

Example declaration:

```json
"launch_handler": {
   "client_mode": " navigate-existing"
}
```

<!-- ---------- -->

> ###### Accept entity info from Windows App Actions framework

Reviewer: ```suggestion
###### Accept entity info from Windows
```

      "packageId": "",
      "url": "https://microsoftedge.github.io/Demos/wami",
      "version": "1.0.2",
      "allowSigning": true,
      "classicPackage": {
         "generate": true,
         "version": "1.0.0"
      },
      "publisher": {
         "displayName": "",
         "commonName": ""
      },
      "edgeChannel": "canary",
      "appUserModelId": "Microsoft.Canary_8wekyb3d8bbwe!MSEDGE",
      "generateModernPackage": true,
      "resourceLanguage": "EN-US",
      "startUrl": "https://microsoftedge.github.io/Demos/wami"
   }
   ```

   Replace the above attributes with your app's actual attributes.

1. Upload your `ActionsManifest.json` file, and then click the **Download** button.

1. Your download is a `.zip` archive that contains files such as `.msix` and `.msixbundle` files.  Unzip it and run `install.ps1` to install your PWA directly.

1. Click the icon for your PWA.

   Your PWA starts.

1. Launch Edge Canary, and then go to `edge://apps`.

   Your PWA is listed.

1. Click the **Details** button.

   Your PWA is indicated as installed from Microsoft Store.

Your PWA has now been installed as a Store PWA in Edge Canary.

<!-- ------------------------------ -->

> #### Testing your PWA in Windows

Reviewer: ```suggestion
#### Test your PWA for App Actions
```

      * [Prerequisites for your PWA](#prerequisites-for-your-pwa)
         * [Web Manifest validation for protocol_handlers](#web-manifest-validation-for-protocol_handlers)
         * [Web Manifest validation for share_target](#web-manifest-validation-for-share_target)
         * [Web Manifest validation for launch_handler](#web-manifest-validation-for-launch_handler)
         * [Accept entity info from Windows App Actions framework](#accept-entity-info-from-windows-app-actions-framework)
   * [Package your PWA](#package-your-pwa)
      * [How to package store PWA in Edge Stable](#how-to-package-store-pwa-in-edge-stable)
      * [How to package store PWA in Edge Canary](#how-to-package-store-pwa-in-edge-canary)
   * [Testing your PWA in Windows](#testing-your-pwa-in-windows)
      * [Using Windows actions test tool](#using-windows-actions-test-tool)

<!-- ====================================================================== -->
## Introduction

Here's how to author and test an app action for the App Actions on Windows framework, for a Progressive Web App (PWA) that you've created.

<!-- ------------------------------ -->
#### What's App Actions on Windows

> An App Action on Windows is an individual unit of behavior that a Windows app can implement and register.  The App Action can then be accessed from other apps and experiences, seamlessly integrating into user workflows.

Reviewer 2: I think we should add a link somewhere in this section like "To learn more about App Actions on Windows, see [link]".

An App Action is an atomic unit of functionality.  An app builds and registers actions, and then Windows or other apps can recommend registered actions to the user at contextually relevant times and locations within the user workflow.

<!-- ------------------------------ -->

> #### Why should you build an Action?

Reviewer 2: Should we consistently say "App Action", or just "Action", or maybe "action" (lowercase)? @novac42 can you help with some branding guidance here?

Reviewer: Then it's better to be "Why should you build actions", actions in lowercase and plural.

`StreamingText` and `RemoteFile` are not yet supported for PWA.  For `File`, `Photo`, `Document` and `Text` entity kinds, see [Entity properties](/windows/ai/app-actions/actions-json#entity-properties) in _Action definition JSON schema for App Actions on Windows_.

<!-- -------------------- -->
##### How does Windows App Actions framework call App Actions in a PWA?

<!-- ---------- -->
###### PWA launch through URI Activation

A PWA action is launched by Windows through custom protocol URI activation.  A PWA must register for custom protocol handling with the `protocol_handlers` field in the web app manifest.  Windows' `LaunchUriAsync(Uri, LauncherOptions, ValueSet)` method is used by the Action Framework for URI activation.

`LaunchUriAsync` doesn't return a value or error status.

<!-- ---------- -->
###### URI scheme

A custom protocol URI should be used, not an HTTPS URI.

For PWAs with many actions, use syntax:

```json

> "uri": "web+{app name://}{custom action name}"

Reviewer 2: ```suggestion
"uri": "web+{app name}://{custom action name}"
```

```

For example:

```json
"uri": "web+wami://paint"
```

<!-- ---------- -->
###### Receiving ValueSet data from Windows and mapping it to Share Target

The activated protocol URI will be mapped to the "url" field of the share target data.

The input `ValueSet` will be converted to a valid JSON string and mapped to the "text" field of the share target data.  That way Edge would determine the protocol handler launch actually represents an Action.

The app's web app manifest must correctly configure the `share_target` field to enable being launched as a share target for an Action.  See [Web Manifest validation for `share_target`](#web-manifest-validation-for-share_target), below.

<!-- -------------------- -->
##### Prerequisites for your PWA

<!-- ---------- -->
###### Web Manifest validation for `protocol_handlers`

> [Required] The `protocol_handlers` field must be present in the manifest, and the protocol field must be the same as the invocation uri in the `ActionsManifest.json` file.

Reviewer 2: The "[Required]" prefix seems a bit weird. How about removing this and adding a table under **Prerequisites for your PWA** that lists all prerequisites, a short description for each, and whether they are required or optional?

Reviewer: ```suggestion
The `protocol_handlers` field must be present in the manifest, and the protocol field must be the same as the invocation uri in the `ActionsManifest.json` file.
```
We can remove the [Required] and [Optional] prefix in the paragraphs. I think at this moment we don't have to add a table, because developers do need to check the detailed config guide and the code snippets in the paragraphs, and the paragraphs are not very lengthy.

> Example `protocol_handlers` declaration:

Reviewer: ```suggestion
Example `protocol_handlers` declaration as below. Please see also [Handle protocols in a PWA](https://learn.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/handle-protocols). 
```

```json
"protocol_handlers": [{
   "protocol": "web+wami",
   "url": "./?url=%s"
}]

> ```

Reviewer 2: Can we add a see also link to how-to/handle-protocols.md?

<!-- ---------- -->
###### Web Manifest validation for `share_target`

> [Required] The `share_target` field must be present in the manifest:

Reviewer: ```suggestion
The `share_target` field must be present in the manifest:
```
Removing [Required]

* The method must be "POST".
* The `enctype` must be `multipart/form-data`.
* `title`, `text` and `url` in `params` are required.  `files` is optional if you don't want your PWA to support the normal _share target_ feature.

Example `share_target` declaration:

```json
"share_target": {
   "action": "./share-target",
   "method": "POST",
   "enctype": "multipart/form-data",
   "params": {
      "title": "title",
      "text": "text",
      "url": "url"
   }
}

> ```

Reviewer 2: Add a link to how-to/share.md#receiving-shared-content

<!-- ---------- -->
###### Web Manifest validation for `launch_handler`

> [Optional] You can set [launch_handler](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/launch_handler) as `navigate-existing` into your PWA manifest, if you want to keep a single PWA window.

Reviewer: ```suggestion
You can set [launch_handler](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/launch_handler) as `navigate-existing` into your PWA manifest, if you want to keep a single PWA window.
```

      "packageId": "",
      "url": "https://microsoftedge.github.io/Demos/wami",
      "version": "1.0.2",
      "allowSigning": true,
      "classicPackage": {
         "generate": true,
         "version": "1.0.0"
      },
      "publisher": {
         "displayName": "",
         "commonName": ""
      },
      "edgeChannel": "canary",
      "appUserModelId": "Microsoft.Canary_8wekyb3d8bbwe!MSEDGE",
      "generateModernPackage": true,
      "resourceLanguage": "EN-US",
      "startUrl": "https://microsoftedge.github.io/Demos/wami"
   }
   ```

   Replace the above attributes with your app's actual attributes.

1. Upload your `ActionsManifest.json` file, and then click the **Download** button.

1. Your download is a `.zip` archive that contains files such as `.msix` and `.msixbundle` files.  Unzip it and run `install.ps1` to install your PWA directly.

1. Click the icon for your PWA.

   Your PWA starts.

1. Launch Edge Canary, and then go to `edge://apps`.

   Your PWA is listed.

1. Click the **Details** button.

   Your PWA is indicated as installed from Microsoft Store.

Your PWA has now been installed as a Store PWA in Edge Canary.

<!-- ------------------------------ -->

> #### Testing your PWA in Windows

Reviewer 2: There's some inconsistencies with how headings are named. Some of the headings are questions, others use gerunds. Can we improve this?

Reviewer: Let's be consistent by removing the question mark and using verb instead of gerunds.

Before testing your PWA, make sure it has been set up and packaged successfully, as discussed in the above sections.

<!-- -------------------- -->

> ##### Using Windows actions test tool

Reviewer 2: This heading seems unnecessary. The **Testing your PWA in Windows** could be renamed to **Test App Actions  in your PWA by using the Windows actions test tool**.

Reviewer: ```suggestion

```
How about: Testing your PWA in Windows -> Test App Actions in your PWA
removing the heading 'Using Windows actions test tool', just show the numbered list.

1. Click the icon for your PWA.

   Your PWA starts.

1. Launch Edge Canary, and then go to `edge://apps`.

   Your PWA is listed.

1. Click the **Details** button.

   Your PWA is indicated as installed from Microsoft Store.

Your PWA was successfully installed as a Store PWA in Edge Canary.

<!-- ------------------------------ -->
#### Testing your PWA in Windows

Before testing, make sure that you have set up and packaged your PWA successfully as discussed in the above sections.

<!-- -------------------- -->
##### Using Windows actions test tool

1. Make sure Windows is updated to support App Actions. Please see [Windows 11 Insider Preview Build 26120.4151 (Beta Channel)](https://blogs.windows.com/windows-insider/2025/05/19/announcing-windows-11-insider-preview-build-26120-4151-beta-channel/) for more information. <!-- link updated -->

1. Make sure Microsoft Edge is updated to the latest version, and install your PWA successfully; see [Package your PWA](#package-your-pwa), above.

1. Launch Edge Canary or Edge Stable, whichever channel you installed the PWA on.

> 1. Go to `edge://flags`, and then enable `#edge-app-actions-on-windows-for-web-apps`.<!-- todo: clarify UI step -->

Author: clarify ui step: edge://flags > enable #edge-app-actions-on-windows-for-web-apps
@novac42

Reviewer: Update to: 
Go to `edge://flags`, search for 'app actions' and locate 'Enable App Actions on Windows for web apps', whose flag name is `#edge-app-actions-on-windows-for-web-apps`. Change the status to Enabled.

          - [Web Manifest validation for `protocol_handlers`](#web-manifest-validation-for-protocol_handlers)
          - [Web Manifest validation for `share_target`](#web-manifest-validation-for-share_target)
          - [Web Manifest validation for `launch_handler`](#web-manifest-validation-for-launch_handler)
          - [Accept entity info from Windows App Actions framework](#accept-entity-info-from-windows-app-actions-framework)
      - [Package your PWA](#package-your-pwa)
        - [How to package store PWA in Edge Stable](#how-to-package-store-pwa-in-edge-stable)
        - [How to package store PWA in Edge Canary](#how-to-package-store-pwa-in-edge-canary)
      - [Testing your PWA in Windows](#testing-your-pwa-in-windows)
        - [Using Windows actions test tool](#using-windows-actions-test-tool)

<!-- ====================================================================== -->
## Introduction

Here's how to author and test an app action for the App Actions on Windows framework, for a Progressive Web App (PWA) that you've created.

<!-- ------------------------------ -->
#### What's App Actions on Windows

> <!-- todo: reused Windows' doc and need link to the public Windows doc -->

Author: link to the public windows doc?
delete this todo?  
@novac42

Reviewer: this todo can be deleted.

The `ActionEntityKind` enumeration specifies the types of entities that are supported by App Actions on Windows.  In the context of a JSON action definition, the entity kinds are string literals that are case-sensitive.  The supported entities are listed in [ActionEntityKind enumeration](/windows/ai/app-actions/actions-json#actionentitykind-enumeration) in _Action definition JSON schema for App Actions on Windows_.

`StreamingText` and `RemoteFile` are not yet supported for PWA.  For `File`, `Photo`, `Document` and `Text` entity kinds, see [Entity properties](/windows/ai/app-actions/actions-json#entity-properties) in _Action definition JSON schema for App Actions on Windows_.

<!-- -------------------- -->
##### How does Windows App Actions framework call App Actions in a PWA?
<!-- updated -->

<!-- ---------- -->
###### PWA launch through URI Activation

A PWA action is launched by Windows through custom protocol URI activation.  A PWA must register for custom protocol handling with the `protocol_handlers` field in the web app manifest.  Windows' `LaunchUriAsync(Uri, LauncherOptions, ValueSet)` method is used by the Action Framework for URI activation.

> `LaunchUriAsync` doesn't yet return a value or error status.<!-- todo -->

Author: Keep this "todo" note, or ok to delete it?  
The wording contains "yet" as a hint that this will need to be updated.  Ideally, in typical articles, omit time cues like that:
https://developers.google.com/style/timeless-documentation
@novac42

Reviewer: good question - Let's revising the wording to remove 'yet'. LaunchUriAsync in the future won't support returning a value either. It will be in other API

You can do either of the following:

* Handle POST share data using server-side code.
* Use a fetch event listener to intercept the HTTP request which allows to access the data in a service worker.  This provides a better experience for offline users.

For action scenario, suggest using service worker to handle POST share data:
* [share_target - Receiving share data using POST](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/share_target#receiving_share_data_using_post)
* [share_target - Receiving shared files](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/share_target#receiving_shared_files)

Here is the example service worker handler: [Support AI Action launch for Wami Web App](https://github.com/MicrosoftEdge/Demos/blob/main/wami/sw.js).<!-- cleaner link -->

The meaning of all `formData` is as follows:

* `formData[<share_target.params.text>]`: All entity info with JSON format (the users' file path will be sanitized).

* `formData[<share_target.params.url>]`: The invocation uri, through which you can know which action this launch came from and handle related parameters if needed.

> * `FormData["windowsActionFiles"]`:<!-- todo: lowercase formData? --> All files' data from action.

Author: ```suggestion
* `formData["windowsActionFiles"]`: All files' data from action.
```
@novac42 - tech review this change

Example `share_target` declaration, from [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json):

```json
"share_target": {
   "action": "./share-target",
   "method": "POST",
   "enctype": "multipart/form-data",
   "params": {
      "title": "title",
      "text": "text",
      "url": "url"
   }
}
```

See also:
* [Receiving shared content](./share.md#receiving-shared-content)

<!-- ====================================================================== -->
## `launch_handler` in the web app manifest

If you want to keep a single PWA window, you can set [launch_handler](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/launch_handler) as `navigate-existing` in your PWA's web app manifest file (such as [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json)).

Example declaration:

```json
"launch_handler": {
   "client_mode": " navigate-existing"
}
```

The above code is not in [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json).

<!-- ====================================================================== -->

> ## Accept entity info on Windows<!-- todo: heading + sentence 1: clarify, what is this section about?  what is "entity info"?  what is the task/accomplishment covered in this section? -->

Author: Clarify what task is covered by section "Accept entity info on Windows"

   * This provides a better experience for offline users.

To use a service worker to handle POST share data:
<!-- todo:
does this apply to both approaches above?
use same wording as one of the above options.  which of the above options is this about?
For the App Action scenario of using a service worker (approach 2?) to handle POST (approach 1?) share data, 
-->
* [share_target - Receiving share data using POST](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/share_target#receiving_share_data_using_post)
* [share_target - Receiving shared files](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/share_target#receiving_shared_files)

The following code is from the wami demo's example service worker handler, to support AI Action launch for the app.  This code is from [Demos/wami/sw.js](https://github.com/MicrosoftEdge/Demos/blob/main/wami/sw.js):

```csharp
// Extract data
const data = {
   title: formData.get('title') || '',
   text: formData.get('text') || '',
   url: formData.get('url') || ''
};
const files = formData.getAll('windowsActionFiles');
```

The above code uses the following `formData`:

<!-- todo

> * **title:** 

Author: should `title` be mentioned in this list?  the other 3 formData lines are listed.
@novac42

-->

> * **text:** `formData[<share_target.params.text>]`: All entity info with JSON format.  The users' file path will be sanitized.<!-- todo: what is formData[<share_target.params.text>], where is that code from?  how does that relate to text: formData.get('text') above? -->

Author: what is formData[<share_target.params.text>], where is that code from?  how does that relate to text: formData.get('text') above?  Is it syntax, for what sort of thing?
@novac42

1. Unzip your download archive.  Then you can [publish your PWA to the Microsoft Store](./microsoft-store.md) and install the app from the Store, or install it directly by running `install.ps1`.

1. Launch your PWA by clicking the PWA icon.

<!-- ====================================================================== -->
## Package a store PWA in Edge Canary

The Stable release of Edge 137 will be available on May 29, 2025; see [Package a store PWA in Edge Stable](#package-a-store-pwa-in-edge-stable), above.

Edge Canary gives a preview of the functionalities that's not available in stable yet.

If you want to try out App Actions for PWAs before May 29, 2025, you can package your PWA in Edge Canary, as follows:

1. Install [Edge Canary](https://www.microsoft.com/edge/download/insider) and upgrade to the latest version.

1. Get the **Package ID**, **Publisher ID**, and **Publisher display name** of your PWA; see [Publish a PWA to the Microsoft Store](./microsoft-store.md).

1. Go to [PWABuilder.Windows.Chromium tester](https://pwabuilder-windows-docker.azurewebsites.net/).

1. Enter your PWA configs in the input box, for the following fields:

   * `name`: Your PWA's name.

>    * `packageId`: Not needed if you won't publish the Canary-bind app<!-- todo: is "Canary-bind app" meaningful? 3x --> to the store; can be an empty string.

Author: is "the Canary-bind app" meaningful?
@captainbrosset @novac42

   * `url`: Your PWA's url.
   * `publisher` > `displayName`: The publisher's name.  Not needed if you won't publish the Canary-bind app to the store; can be an empty string.
   * `publisher` > `commonName`: Your Publisher ID.  Not needed if you won't publish the Canary-bind app to the store; can be an empty string.
   * `startUrl`: Your PWA's `startUrl`.

>    Example of a JSON listing:<!-- todo: which file?  where is this snippet from?  need a lead-in, to introduce each code snippet and give context: --><!-- todo: is this the complete file/ JSON object? -->

Author: Where is this listing from?  Generated as you fill in the form?  Is this an example app actions definition manifest file such as ActionsManifest.json but specialized for Canary?
Above each code snippet, need a lead-in giving file name; type of file; context.
@novac42

---
title: Enable App Actions on Windows for a PWA
description: How to author and test an App Action for the App Actions on Windows framework, for a Progressive Web App (PWA) that you've created.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.service: microsoft-edge
ms.subservice: pwa
ms.date: 05/19/2025
---
# Enable App Actions on Windows for a PWA

You can author and test an App Action for the App Actions on Windows framework, for a Progressive Web App (PWA) that you've created.  An App Action on Windows is an individual unit of behavior that a Windows app can implement and register.  The App Action can then be accessed from other apps and experiences, seamlessly integrating into user workflows.

> An app builds and registers each App Action, and then Windows apps or non-Windows apps can recommend registered App Actions to the user at contextually relevant times and locations within the user's workflow.

Reviewer 2: "Windows apps or non-Windows apps" feels a bit odd to me. What's a non-Windows app?
What about this instead:

```suggestion
An app builds and registers each App Action, and then Windows apps and experiences can recommend registered App Actions to the user at contextually relevant times and locations within the user's workflow.
```

> You can build Windows App Actions to increase discoverability and engagement of your app's features by using app-to-app and agentic interactions.  The Windows App Actions framework makes it possible to integrate your app with other apps or with AI agents.

Reviewer 2: The App Actions on Windows docs say this: "App Actions on Windows are individual units of behavior that a Windows app can implement and register so that they can be accessed from other apps and experiences", and never mention the word AI agent or agentic. So I don't think we should mention agents.

How about we simplify this to:

```suggestion
You can build Windows App Actions to increase discoverability and engagement of your app's features.
```

See also:

> * [App Actions on Windows Overview](/windows/ai/app-actions/)

Reviewer: If we use relative path, edge-developer does not contain the path windows/ai/app-actions. The Windows doc link is in https://learn.microsoft.com/en-us/windows/ai/app-actions/

Author: the prefix-trimmed line 20 is correct; partial path + LMC prefix, end up at:
https://learn.microsoft.com/en-us/windows/ai/app-actions/

**Detailed contents:**

> * [Recommended scenarios](#recommended-scenarios)

Reviewer: relative path issue. link is under https://learn.microsoft.com/en-us/windows/ai/app-actions/

> * [Define App Actions in the action definition manifest](#define-app-actions-in-the-action-definition-manifest)

Reviewer: relative path issue. link is under https://learn.microsoft.com/en-us/windows/ai/app-actions/ with the "#"tag

Author: this seems a stray comment

>    * [Properties supported in an action definition manifest](#properties-supported-in-an-action-definition-manifest)

Reviewer: relative path issue. link is under https://learn.microsoft.com/en-us/windows/ai/app-actions/ with the "#"tag

   * [Supported Entity types for App Actions for PWAs](#supported-entity-types-for-app-actions-for-pwas)
* [How a PWA is launched through URI activation](#how-a-pwa-is-launched-through-uri-activation)
   * [uri in the action definition manifest](#uri-in-the-action-definition-manifest)
   * [Receive ValueSet data and map it to a share target](#receive-valueset-data-and-map-it-to-a-share-target)
* [Prerequisites for your PWA to support app actions](#prerequisites-for-your-pwa-to-support-app-actions)
   * [share_target in the web app manifest](#share_target-in-the-web-app-manifest)
   * [protocol_handlers in the web app manifest](#protocol_handlers-in-the-web-app-manifest)
   * [launch_handler in the web app manifest](#launch_handler-in-the-web-app-manifest)
   * [Handle the entity of an app action](#handle-the-entity-of-an-app-action)
* [Package a store PWA in Edge Stable](#package-a-store-pwa-in-edge-stable)
* [Package a store PWA in Edge Canary](#package-a-store-pwa-in-edge-canary)
* [Test App Actions for a PWA](#test-app-actions-for-a-pwa)
* [See also](#see-also)

<!-- ------------------------------ -->
#### Recommended scenarios

> The Windows App Actions framework is for exposing and executing key pieces of app functionality, not generic user activities such as opening a right-click menu (context menu).  See [Recommended scenarios for App Actions](/windows/ai/app-actions#recommended-scenarios-for-app-actions) in _App Actions on Windows Overview_.

Author: /windows/ai/app-actions#recommended-scenarios-for-app-actions
+
https://learn.microsoft.com
= 
https://learn.microsoft.com/windows/ai/app-actions#recommended-scenarios-for-app-actions

ends up at:
https://learn.microsoft.com/en-us/windows/ai/app-actions/#recommended-scenarios-for-app-actions

Also test rendered page eg 
https://github.com/mikehoffms/edge-developer/blob/user/mikehoffms/branch-02/microsoft-edge/progressive-web-apps-chromium/how-to/app-actions.md

```json
"uri": "web+{app name}://{custom action name}"
```

For example:

```json
"uri": "web+wami://paint"
```

The above line is shown in context in [Define App Actions in the action definition manifest](#define-app-actions-in-the-action-definition-manifest), above.

<!-- ====================================================================== -->
#### Receive `ValueSet` data and map it to a share target

The activated protocol URI will be mapped to the `url` field of the share target data.

The input `ValueSet` will be converted to a valid JSON string and mapped to the `text` field of the share target data.  That enables Microsoft Edge to determine that the protocol handler launch represents an App Action.

The app's web app manifest (such as [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json)) must correctly configure the `share_target` field, to enable being launched as a share target for an App Action.  See [`share_target` in the web app manifest](#share_target-in-the-web-app-manifest), below.

<!-- ====================================================================== -->
## Prerequisites for your PWA to support app actions

> #### `share_target` in the web app manifest

Reviewer 2: Maybe add a short intro at the top of this h2 heading, to separate it from the first h4.

```suggestion
To correctly integrate your PWA with the App Actions on Windows framework, the following web app manifest members are required:

* `share_target`
* `protocol_handlers`
* `launch_handler`

#### `share_target` in the web app manifest
```

   "protocol": "web+wami",
   "url": "./?url=%s"
}]
```

The matching `invocation` > `uri` field from `ActionsManifest.json`:

```json
"invocation": {
   "type": "Uri",
   "uri": "web+wami://paint"
},
```

The above snippet is shown in context in [Define App Actions in the action definition manifest](#define-app-actions-in-the-action-definition-manifest), above.

See also:
* [Handle protocols in a PWA](./handle-protocols.md)

<!-- ====================================================================== -->
#### `launch_handler` in the web app manifest

If you want to keep a single PWA window, you can set [launch_handler](https://developer.mozilla.org/docs/Web/Progressive_web_apps/Manifest/Reference/launch_handler) as `navigate-existing` in your PWA's web app manifest file (such as [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json)).

Example declaration:

```json
"launch_handler": {
   "client_mode": " navigate-existing"
}
```

The above code is not in [Demos/wami/manifest.json](https://github.com/MicrosoftEdge/Demos/blob/main/wami/manifest.json).

<!-- ====================================================================== -->

> #### Handle the entity of an app action

Reviewer 2: This last h4 section here is not about using a web app manifest member, so it's a bit inconsistent with the previous 3 h4 sections. Also, it seems to me like what's described here is the logic that's necessary to handle `share_target` correctly. 
How about we move this to be inside the **`share_target` in the web app manifest** h4, as a new h6?
